### PR TITLE
Add Backoff/ExponentialBackoff

### PR DESCRIPTION
**Pull request by mmihic**
Adds a stateful mechanism for computing backoffs, with a specific implementation for computing exponential backoffs with optional jitter.

### CODE REVIEW

File: misk/src/main/kotlin/misk/backoff/ExponentialBackoff.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
+package misk.backoff
+
+import java.time.Duration
+import java.util.concurrent.ThreadLocalRandom
+
+/**
+ * Performs exponential backoff with optional jitter. Durations are supplied as
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/backoff/ExponentialBackoff.kt]: why an optional jitter? or at least make it the default

File: misk/src/main/kotlin/misk/backoff/ExponentialBackoff.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  private val baseDelay: () -> Duration,
+  private val maxDelay: () -> Duration,
+  private val jitter: () -> Duration // TODO(mmihic): Consider making jitter a %
+
+) : Backoff {
+  private var consecutiveRetryCount: Int = 0
+  private var maxRetryCount = Integer.MAX_VALUE
+
+  /**
+   * Creates a new unjittered [ExponentialBackoff] using a function for the base
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/backoff/ExponentialBackoff.kt]: nit: not a function, a supplier

File: misk/src/main/kotlin/misk/backoff/ExponentialBackoff.kt [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+   */
+  constructor(baseDelay: Duration, maxDelay: Duration)
+      : this(baseDelay, maxDelay, Duration.ofMillis(0))
+
+  override fun reset() {
+    consecutiveRetryCount = 0
+  }
+
+  override fun nextRetry(): Duration {
+    // Cap the retry count to prevent long overflow. Since it's a function of the configurable base
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/backoff/ExponentialBackoff.kt]: :)

File: misk/src/main/kotlin/misk/backoff/ExponentialBackoff.kt (hunk 2 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/backoff/ExponentialBackoff.kt]: It's a function not a supplier

File: misk/src/main/kotlin/misk/backoff/ExponentialBackoff.kt (hunk 1 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/backoff/ExponentialBackoff.kt]: What's a good default (given that we don't know any either the base or the max delay)?

Comment (ryanhall07) [reply misk/src/main/kotlin/misk/backoff/ExponentialBackoff.kt]: Maybe a default percentage?

Comment (mmihic) [reply misk/src/main/kotlin/misk/backoff/ExponentialBackoff.kt]: You ok if I deal with this in another PR? I have a higher priority branch that's using backoffs that I want to get in first.

Comment (ryanhall07) [reply misk/src/main/kotlin/misk/backoff/ExponentialBackoff.kt]: definitely. just a suggestion

**Pull request merged**